lib/SILGen/SILGenStmt.cpp

```diff
@@ -1071,9 +1078,27 @@ void StmtEmitter::visitForEachStmt(ForEachStmt *S) {
                 S, inputValue, optTL, SGFContext(initLoopVars.get()));
           }
 
-          if (!inputValue.isInContext())
-            RValue(SGF, S, optTy.getOptionalObjectType(), inputValue)
-                .forwardInto(SGF, S, initLoopVars.get());
+          if (!inputValue.isInContext()) {
+            auto initSrc =
+                RValue(SGF, S, optTy.getOptionalObjectType(), inputValue);
+            if (inputValue.getOwnershipKind() ==
+                ValueOwnershipKind::Guaranteed) {
+              std::move(initSrc).copyInto(SGF, S, initLoopVars.get());
```

Comment:
Copying the value seems contrary to the goals of implementing a __shared specifier on the element binding. Is an optimizer pass going to always eliminate the copy, or is this a transitory step towards copy-free iteration?

Reply:
It's a good question. The reason I copied it instead of forwarding is because I ran into the assertion `isPlusOne(SGF)`  in `forwardInto`. When I removed the assertion, cleanup, and extended the borrow scope, both the copy and borrow scope were eliminated. I'll see if I can find any problems forwarding borrowed values because that would be the ideal solution. 

For what it's worth, the optimizer should easily be able to clean up the pattern:
```
x : @guaranteed
y = copy_value x
begin_borrow y
```

Reply:
My understanding is that forwarding means an ownership transfer, so it doesn't make sense to forward a borrowed value since you don't own it. It should be possible to just pass the borrowed value down into the body without forwarding or copying it, no? But maybe I'm totally wrong. CC @gottesmm

Reply:
I don't know about the terminology as far as what "forwarding" means. I'll take your word for it :) The current implementation uses an "Initialization" but I'm not sure that's needed (at least in some cases) for guaranteed values. I'll see about circumnavigating the initialization and just passing the borrowed argument down. 

Maybe forwarding isn't the right name but, with these changes forwarding a guaranteed values is essentially a no-op.